tests/test_data.py

```diff
@@ -196,3 +198,198 @@ def test_zero_length_fails():
+    """Check that DictDataset preserves array dtype."""
+    for _ in range(n_checks):
+        n_samples = np.random.randint(max_batch_size) + 1
+        sample = dict_dataset.sample(n_samples)
+        for k in data_map.keys():
+            assert data_map[k].dtype == sample[k].dtype
+
+
+@pytest.mark.parametrize("n_samples", [-i for i in range(4)])
+def test_dict_dataset_nonpositive_samples_error(dict_dataset, n_samples):
+    """Check that DictDataset errors on n_samples<=0"""
+    with pytest.raises(ValueError, match="n_samples"):
+        dict_dataset.sample(n_samples)
+
+
+def test_dict_dataset_data_map_error(dict_dataset_params):
+    """Check that DictDataset errors on unequal number of data_map rows."""
+    dataset_cls, kwargs = dict_dataset_params
+    with pytest.raises(ValueError, match="Empty.*"):
+        dataset_cls({}, **kwargs)
+
+    unequal_data_map = {
+        "a": np.zeros([10]),
+        "b": np.zeros([10, 1]),
+        "c": np.zeros([10, 2]),
+        "d": np.zeros([2, 10]),
+    }
+    with pytest.raises(ValueError, match="Unequal.*"):
+        dataset_cls(unequal_data_map, **kwargs)
+
+
+def test_dict_dataset_parallel_rows(
+    dict_dataset_params, max_batch_size=80, n_checks=20
+):
+    """Check that 'parallel rows' among different key-value pairs remain parallel.
+
+    Nontrivially, shuffled datasets should maintain this order.
+    """
+    dataset_cls, kwargs = dict_dataset_params
+    range_data_map = {k: np.arange(50,) for k in "abcd"}
+    dict_dataset = dataset_cls(range_data_map, **kwargs)
+    for _ in range(n_checks):
+        n_samples = np.random.randint(max_batch_size) + 1
+        sample = dict_dataset.sample(n_samples)
+        for v in sample.values():
+            np.testing.assert_array_equal(sample["a"], v)
+
+
+def test_dict_dataset_correct_sample_len(dict_dataset, max_batch_size=80):
+    """Check that DictDataset returns samples of the right length."""
+    batch_sizes = np.arange(max_batch_size) + 1
+    np.random.shuffle(batch_sizes)
+    for n_samples in batch_sizes:
+        sample = dict_dataset.sample(n_samples)
+        for v in sample.values():
+            assert len(v) == n_samples
+
+
+@pytest.mark.parametrize("shuffle", [False, True])
+@pytest.mark.parametrize("dataset_size", [1, 30, 100, 200])
+class TestEpochOrderDictDataset:
+    @pytest.fixture
+    def arange_dataset(self, shuffle, dataset_size):
+        data_map = {"a": np.arange(dataset_size)}
+        ds = dataset.EpochOrderDictDataset(data_map, shuffle=shuffle)
+        return ds
+
+    def test_epoch_order_dict_dataset_shuffle_order(
+        self, arange_dataset, shuffle, dataset_size, n_checks=3,
+    ):
+        """Check that epoch order is deterministic iff not shuffled.
+
+        The check has `factorial(dataset_size)` chance of false negative when shuffle
+        is True, so we skip on smaller dataset_sizes.
+        """
+        if dataset_size < 100 and shuffle:
```

Comment:
100! is pretty large, 10! is already larger than 1e6. If you're sure it's factorial, then I think you can reduce the threshold.

Reply:
Alright, I'll change to 20!